crates/devices/virtio-console/src/console.rs

```diff
@@ -0,0 +1,176 @@
+use std::sync::{Arc, Mutex};
+use virtio_queue::DescriptorChain;
+use vm_memory::{Bytes, GuestMemory, GuestMemoryError};
+
+/// Virtio console chain processing abstraction.
+///
+/// This module offers an abstraction for working with the descriptor chains of the receive and
+/// transmit queues.
+///  * The transmit queue is used to send data from the driver to the device and it's buffers are
+///    read-only for the device.
+///  * The receive queue is used to send data from the device to the driver and it's buffers are
+///    write-only for the device.
+///
+/// To implement this abstraction the [`Console`](struct.Console.html)  structure is provided which
+/// consists of the `input_buffer` member and exposes several methods.
+/// * The [`input_buffer`](struct.Console.html#field.input_buffer) is a buffer used to store data
+///   before sending it to the driver. The following methods are provided to interact with it:
+///   - [`enqueue_data`](struct.Console.html#method.enqueue_data) which is used to add data to the
+///     end of the buffer.
+///   - [`is_input_buffer_empty`](struct.Console.html#method.is_input_buffer_empty) which is used
+///     assess if there is any data waiting to be sent.
+///
+/// The methods for processing a descriptor chain are the following:
+/// *  [`process_transmitq_chain`](struct.Console.html#method.process_transmitq_chain) which is
+///    used to read the data from a descriptor chain of the transmit queue.
+/// * [`process_receiveq_chain`](struct.Console.html#method.process_receiveq_chain) which is used
+///   to send the data to the driver by writing it into the buffers of a receive queue descriptor
+///   chain.
+
+/// Console device errors.
+#[derive(Debug)]
+pub enum Error {
+    /// Failed to write slice to guest memory.
+    WriteToGuestFailed(GuestMemoryError),
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match self {
+            WriteToGuestFailed(ref err) => {
+                write!(f, "console failed to write slice to guest memory: {}", err)
+            }
+        }
+    }
+}
+
+/// Console struct that implements the abstraction of virtio console descriptor chain handling.
+pub struct Console {
+    /// Buffer that stores data to be sent to the driver.
+    input_buffer: Arc<Mutex<Vec<u8>>>,
+}
+
+impl Console {
+    /// Create new console object.
+    pub fn new() -> Self {
+        Console {
+            input_buffer: Arc::new(Mutex::new(Vec::new())),
+        }
+    }
+
+    /// Adds data in the form of raw bytes to the input buffer.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - A mutable reference to a `Vec<u8>` containing the raw bytes that will be
+    ///            sent to the driver when the `process_receiveq_chian` method is called.
+    pub fn enqueue_data(&self, data: &mut Vec<u8>) {
+        let mut input_buffer = self.input_buffer.lock().unwrap();
+        input_buffer.append(data);
```

Comment:
We should probably add a check here to not allow unlimited growth of the input buffer.